### ISSUE TITLE
Unable to incorporate SignalR because of internalized classes

### ISSUE THREAD

**Issue by thoemmi** [aspnet/SignalR#2233]
I'm trying to add SignalR (1.0.0-rc1-final) into our application. This application is a backend server (a monolithic Windows service), which among other things provides a web API. We're using Autofac because Microsoft's dependency injection does not support nested containers.

That's the reason why calling `services.AddSignalR()` is not an option for us. The singletons must be registered in the root container, outside the scope of the web API.
When we started with *1.0.0-preview1-final*, we inspected the codepath of `services.AddSignalR()` and registered all required services in our own Autofac container, so other components could send messages to SignalR clients.

However, when upgrading to *1.0.0-rc1-final* we noticed that this is not possible anymore, as you have changed the visibility of two classes to internal, namely `HubContext<THub>` and `HubContext<THub, T>`.

Additionally you have added the class `SignalRMarkerService` as a "service guard", introduced in #1562. Because it's internal too, I was forced to use reflection to register that class in Autofac:

```csharp
var assembly = Assembly.Load("Microsoft.AspNetCore.SignalR.Core");
var signalRMarkerServiceType = assembly.GetType("Microsoft.AspNetCore.SignalR.Internal.SignalRMarkerService", true);
builder.RegisterType(signalRMarkerServiceType).SingleInstance();
```

So my request is to make the three relevant classes (`HubContext<THub>`, `HubContext<THub, T>`, and `SignalRMarkerService`) public (again).

**Comment by davidfowl** [comment]
> I'm trying to add SignalR (1.0.0-rc1-final) into our application. This application is a backend server (a monolithic Windows service), which among other things provides a web API. We're using Autofac because Microsoft's dependency injection does not support nested containers.

I don't really understand why you're not using the built in DI. Can you elaborate more on how you are using SignalR, we have no plans to make those types public.

**Comment by pielegacy** [comment]
I second @thoemmi, my team's project relied on `HubContext<THub>` and `HubContext<THub, T>` as we were extending the functionality of the context to provide an additional group of clients that messages could be sent to which was automatically inferred from the current `HttpContext`. With these changes (which were very unexpected for a release candidate) that entire system no longer works as the required types are no longer exposed.

**Comment by davidfowl** [comment]
@pielegacy 

> I second @thoemmi, my team's project relied on HubContext<THub> and HubContext<THub, T> as we were extending the functionality of the context to provide an additional group of clients that messages could be sent to which was automatically inferred from the current HttpContext

Why can't you use the interface and mimic the default implementation? I'm scared to ask how you were using the HubContext with the IHttpContextAccessor since it's a recipe for disaster if you use it anywhere outside of an http request. For example, it'll break if you're using long polling and are in a hub method (and it's a race condition so it won't repro all of the time).

If you're desperate, you can copy the 16 lines of actual code here: https://github.com/aspnet/SignalR/blob/023e61495f1ab2bd59efc6d3582f59d46f6fdcbf/src/Microsoft.AspNetCore.SignalR.Core/Internal/HubContext.cs

> With these changes (which were very unexpected for a release candidate) that entire system no longer works as the required types are no longer exposed.

    As ASP.NET Core SignalR is pre-release software, al bets are off until we ship 1.0. We try out best 
    not to break from rc1 to 1.0.0 but before that we make changes are our discretion.

**Comment by pielegacy** [comment]
Thanks for the response, will take your recommendations into account and ensure were not using Http context unsafely

**Comment by thoemmi** [comment]
Our (monolithic) application is divided into a couple of modules. One of those modules provides a web API and also notifications using SignalR.
At application startup the main Autofac container is built and afterwards all classes implementing a specific interface are called to start. One of those classes will then create the web API host using the `WebHostBuilder`. Because the root container is already in read-only state, we create a child container. This also isolates the MVC registrations from the main container.
However, other components (outside the web API scope) will generate notifications, which should be forwarded to SignalR clients too, therefore we need the `HubLifetimeManager<>` instance at the root level.

TL;DR: we use nested Autofac containers and need to register certain SignalR services one level above the web API container.

**Comment by davidfowl** [comment]
I'd love to see a code sample so get a better understanding of what the application even looks like. Unfortunately at this stage, we're not making any API changes so you'll need to keep the private reflection hack for now. As for the HubContext classes, they are trivial to copy so hopefully that won't be too much of a burden.

**Comment by thoemmi** [comment]
@davidfowl Ok, I'll try to strip down our application to the simplest form.

**Comment by thoemmi** [comment]
Hi @davidfowl, I've tried to create a repro as simple as possible, please see https://github.com/thoemmi/NestedContainersAndSignalR. It includes a readme, giving a rough overview of the architecture. My concern is the [quite lengthy registration of SignalR services](https://github.com/thoemmi/NestedContainersAndSignalR/blob/master/src/Sample.Module.WebApi/WebApiModule.cs#L31-L51), especially using reflaction go get access to the internal classes mentioned above.

**Comment by thoemmi** [comment]
Thanks for having a look at our architecture.

> Do those other modules run inside of the web application or outside (otherwise it wouldn't work unless you were using redis or another backplane)?

Actually, all modules are registered at root level. Only the web host gets it's own child container because of its special requirement to own the container.

But why's SignalR so restrictive? Why do you make it so hard to add SignalR to our root container, so that we have to use reflection? It would help use if `HubContext` would be public and SignalR wouldn't use this service guard `SignalRMarkerService` introduced in #1562. 

Generally, IMHO it's sad that self-hosting of webhost is not a first-class citizen.

**Comment by davidfowl** [comment]
> But why's SignalR so restrictive? Why do you make it so hard to add SignalR to our root container, so that we have to use reflection? It would help use if HubContext would be public and SignalR wouldn't use this service guard SignalRMarkerService introduced in #1562.

SignalR isn't restrictive, it works like every other ASP.NET Core framework. It has an extension method on the IServiceCollection to prepare the services it requires to function. Some of those services are internal (by design). The way to make this work is to use that mechanism to flow services to into your root container. What if we add another required service in the next version? You're setup will break.

> Generally, IMHO it's sad that self-hosting of webhost is not a first-class citizen.

I don't really have a comment on this. If you can think of a way to make this be reasonable, I'm all ears but it fundamentally does not fit with the existing model. The entire premise around adding services to a service collection wouldn't work. How would MVC and SignalR get services into your autofac module? Do we write bindings for ever framework? We fixed that problem by making our extensions hang off IServiceCollection. It's the lingua franca for service registration for us.

**Comment by thoemmi** [comment]
I do see your point. However, I also see that others have similar requirements (e.g. https://github.com/aspnet/Hosting/issues/1349 or https://github.com/dotnet/orleans/issues/4236). Therefore I'll close this issue. Thanks for taking your time.
